package frc.robot.subsystems.intake;

import static org.wpilib.units.Units.Degrees;
import static org.wpilib.units.Units.Hertz;
import static org.wpilib.units.Units.RPM;
import static org.wpilib.units.Units.Radians;

import coppercore.controls.state_machine.StateMachine;
import coppercore.monitors.TotalCurrentCalculator;
import coppercore.parameter_tools.LoggedTunableNumber;
import coppercore.wpilib_interface.MonitoredSubsystem;
import coppercore.wpilib_interface.subsystems.motors.MotorIO;
import coppercore.wpilib_interface.subsystems.motors.MotorInputsAutoLogged;
import coppercore.wpilib_interface.tuning.TestModeManager;
import org.wpilib.math.util.Units;
import org.wpilib.units.measure.Angle;
import org.wpilib.units.measure.AngularVelocity;
import org.wpilib.units.measure.Voltage;
import org.wpilib.driverstation.internal.DriverStationBackend;
import org.wpilib.system.RobotController;
import frc.robot.constants.JsonConstants;
import frc.robot.util.StateMachineDump;
import java.util.List;
import org.littletonrobotics.junction.Logger;

public class IntakeSubsystem extends MonitoredSubsystem {

  TestModeManager<PivotTestMode> pivotTestModeManager =
      new TestModeManager<PivotTestMode>("IntakePivot", PivotTestMode.class);

  TestModeManager<RollerTestMode> rollerTestModeManager =
      new TestModeManager<RollerTestMode>("IntakeRollers", RollerTestMode.class);

  StateMachine<IntakeSubsystem> intakeStateMachine;

  protected MotorIO pivotMotorIO;
  protected MotorIO rollersLeadMotorIO;
  protected MotorIO rollersFollowerMotorIO;

  protected MotorInputsAutoLogged pivotInputs;
  protected MotorInputsAutoLogged rollerLeadMotorInputs;
  protected MotorInputsAutoLogged rollerFollowerMotorInputs;

  protected Angle targetPivotAngle = Degrees.zero();
  protected Voltage holdVoltage = null;

  private LoggedTunableNumber rollersTargetSpeedTunable;

  private IntakeDependencies dependencies = new IntakeDependencies();

  // Dependencies (these are what we would have fetched using extensive supplier networks in 2025
  // and before)
  public static class IntakeDependencies {
    /**
     * Whether or not the homing switch is currently pressed. This value should default to false
     * when a homing limit switch is not present.
     */
    private boolean isHomingSwitchPressed = false;

    public boolean isHomingSwitchPressed() {
      return isHomingSwitchPressed;
    }
  }

  public IntakeDependencies getDependencies() {
    return this.dependencies;
  }

  public void setIsHomingSwitchPressed(boolean isHomingSwitchPressed) {
    dependencies.isHomingSwitchPressed = isHomingSwitchPressed;
  }

  public IntakeSubsystem(
      MotorIO pivotMotorIO, MotorIO rollersLeadMotorIO, MotorIO rollersFollowerMotorIO) {
    this.pivotMotorIO = pivotMotorIO;
    this.rollersLeadMotorIO = rollersLeadMotorIO;
    this.rollersFollowerMotorIO = rollersFollowerMotorIO;

    this.pivotInputs = new MotorInputsAutoLogged();
    this.rollerLeadMotorInputs = new MotorInputsAutoLogged();
    this.rollerFollowerMotorInputs = new MotorInputsAutoLogged();

    this.rollersTargetSpeedTunable =
        new LoggedTunableNumber(
            "IntakeTunables/RollersTargetSpeedRPM",
            JsonConstants.intakeConstants.intakeTeleOpRollerSpeed.in(RPM));

    this.intakeStateMachine = new StateMachine<IntakeSubsystem>(this);

    IntakeState.testModeState =
        this.intakeStateMachine.registerState(new IntakeState.TestModeState(this));
    List.of(
            IntakeState.controlToPositionState,
            IntakeState.waitForButtonState,
            IntakeState.homingWaitForMovementState,
            IntakeState.homingWaitForStopMovingState,
            IntakeState.homingDoneState)
        .forEach(this.intakeStateMachine::registerState);

    pivotMotorIO.setRequestUpdateFrequency(Hertz.of(1000));

    // ### Test Mode Transitions
    // Any time we finish any state and we should be in test mode, we transition to the test mode
    // state at
    // the soonest time we can without disrupting the homing process. So that when we are in
    // test mode,
    // we can be sure that it has been properly homed and that the setpoints we get in test mode are
    // accurate.
    IntakeState.waitForButtonState
        .whenFinished()
        .andWhen(IntakeState::shouldBeInTestMode, "Should be in test mode")
        .transitionTo(IntakeState.testModeState);
    IntakeState.homingDoneState
        .whenFinished()
        .andWhen(IntakeState::shouldBeInTestMode, "Should be in test mode")
        .transitionTo(IntakeState.testModeState);
    IntakeState.controlToPositionState
        .when(IntakeState::shouldBeInTestMode, "Should be in test mode")
        .transitionTo(IntakeState.testModeState);
    IntakeState.controlToPositionState.whenRequestedTransitionTo(
        IntakeState.homingWaitForMovementState);
    IntakeState.testModeState.whenFinished().transitionTo(IntakeState.controlToPositionState);

    // ### Homing Button Transitions
    IntakeState.waitForButtonState.whenFinished().transitionTo(IntakeState.controlToPositionState);
    IntakeState.waitForButtonState
        .when(DriverStationBackend::isEnabled, "When robot is enabled and button has not been pressed")
        .transitionTo(IntakeState.homingWaitForMovementState);

    // ### Wait for movement transitions
    // If the robot gets disabled during the homing process, we transition back to
    // the wait for button state to wait for the operator to re-enable the robot
    //  and restart the homing process.
    IntakeState.homingWaitForMovementState
        .when(DriverStationBackend::isDisabled, "When robot is disabled during homing")
        .transitionTo(IntakeState.waitForButtonState);
    // If the mechanism starts moving, we assume that it is has started the homing
    // process properly and we transition to the homing wait for stop moving state
    // to wait for it stop moving to finish the homing process
    IntakeState.homingWaitForMovementState
        .whenFinished()
        .transitionTo(IntakeState.homingWaitForStopMovingState);
    // If the mechanism doesn't start moving within the timeout, we assume that it
    // is already in the homed position and we finish the homing process by
    // transitioning to the homing done state
    IntakeState.homingWaitForMovementState
        .whenTimeout(JsonConstants.intakeConstants.homingTimeoutSeconds)
        .transitionTo(IntakeState.homingDoneState);

    // If the robot gets disabled during the homing process, we transition back to
    // the wait for button state to wait for the operator to re-enable the robot
    //  and restart the homing process.
    IntakeState.homingWaitForStopMovingState
        .when(DriverStationBackend::isDisabled, "When robot is disabled during homing")
        .transitionTo(IntakeState.waitForButtonState);
    // If the mechanism starts moving, we assume that we have started the homing
    // process properly and so we wait for it to stop moving by hitting a hard
    // stop. Once it stops moving, we assume that we are in the homed position
    // and we finish the homing process by transitioning to the homing done state
    IntakeState.homingWaitForStopMovingState
        .whenFinished()
        .transitionTo(IntakeState.homingDoneState);

    // ### Exiting homing process transition
    // Once the mechanism has stopped moving, we consider the homing process to be
    // done and we set the current position as zero and transition to the control
    // to position state
    IntakeState.homingDoneState.whenFinished().transitionTo(IntakeState.controlToPositionState);

    this.intakeStateMachine.setState(IntakeState.waitForButtonState);
    StateMachineDump.write("intake", this.intakeStateMachine);
  }

  public void runRollers(AngularVelocity rollerSpeed) {
    rollersLeadMotorIO.controlToVelocityProfiled(rollerSpeed);
  }

  public void stopRollers() {
    rollersLeadMotorIO.controlNeutral();
    // We don't need to command the follower motor to stop because
    // it is always following the lead motor
  }

  public void setTargetPivotAngle(Angle angle) {
    this.targetPivotAngle = angle;
    this.holdVoltage = null;
  }

  public Angle getCurrentTargetPivotAngle() {
    return this.targetPivotAngle;
  }

  public Angle getCurrentPivotAngle() {
    return Radians.of(this.pivotInputs.positionRadians);
  }

  // Should these be blocked from executing if we are in test mode?

  public void stow() {
    setTargetPivotAngle(JsonConstants.intakeConstants.stowPositionAngle);
    // needsReHome = true;
  }

  public void deploy() {
    setTargetPivotAngle(JsonConstants.intakeConstants.intakePositionAngle);
    holdVoltage = JsonConstants.intakeConstants.pivotVoltageWhenIntaking;
    // if (needsReHome) {
    //   //intakeStateMachine.requestState(IntakeState.homingWaitForMovementState);
    //   needsReHome = false;
    // }
  }

  public void applyHoldVoltage() {
    pivotMotorIO.controlOpenLoopVoltage(holdVoltage);
  }

  @Override
  public void monitoredPeriodic() {
    long startTimeUs = RobotController.getTime();

    pivotMotorIO.updateInputs(pivotInputs);
    rollersLeadMotorIO.updateInputs(rollerLeadMotorInputs);
    rollersFollowerMotorIO.updateInputs(rollerFollowerMotorInputs);

    TotalCurrentCalculator.recordCurrent(
        hashCode(),
        pivotInputs.supplyCurrentAmps
            + rollerLeadMotorInputs.supplyCurrentAmps
            + rollerFollowerMotorInputs.supplyCurrentAmps);

    Logger.processInputs("Intake/Pivot/Inputs", pivotInputs);
    Logger.processInputs("Intake/RollerLead/Inputs", rollerLeadMotorInputs);
    Logger.processInputs("Intake/RollerFollower/Inputs", rollerFollowerMotorInputs);

    // This is run outside of the test mode because we want to be able to tune the roller speed
    // in test mode and be able to control the pivot as if we were operating the robot normally
    if (rollerTestModeManager.getTestMode() == RollerTestMode.RollerSpeedTuning) {
      LoggedTunableNumber.ifChanged(
          hashCode(),
          rollerSpeed ->
              JsonConstants.intakeConstants.intakeTeleOpRollerSpeed = RPM.of(rollerSpeed[0]),
          rollersTargetSpeedTunable);
    }

    Logger.recordOutput("Intake/State", intakeStateMachine.getCurrentState().getName());

    intakeStateMachine.periodic();

    // Ensure that even if we accidentally command the follower motor to do something
    // it won't cause any issues because we always command it to follow the lead motor
    // at the end of the periodic
    rollersFollowerMotorIO.follow(JsonConstants.canBusAssignment.intakeRollersLeadMotorId, false);

    long endTimeUs = RobotController.getTime();
    if (JsonConstants.featureFlags.logPeriodicTiming) {
      Logger.recordOutput("PeriodicTime/IntakeMs", (endTimeUs - startTimeUs) / 1000.0);
    }
  }

  protected void controlToTargetPivotAngle() {
    pivotMotorIO.controlToPositionUnprofiled(this.targetPivotAngle);
  }

  public void controlPivotMotorIOWithVoltage(Voltage v) {
    pivotMotorIO.controlOpenLoopVoltage(v);
  }

  protected void setPositionIfOutsideRange() {
    // Commented out because the intake can actually go down to ~-8 degrees now.
    if (pivotInputs.positionRadians
        < JsonConstants.intakeConstants.minPivotAngle.in(Radians) - Units.degreesToRadians(0.5)) {
      pivotMotorIO.setCurrentPosition(JsonConstants.intakeConstants.minPivotAngle);
    } else if (pivotInputs.positionRadians
        > JsonConstants.intakeConstants.maxPivotAngle.in(Radians) + Units.degreesToRadians(0.5)) {
      pivotMotorIO.setCurrentPosition(JsonConstants.intakeConstants.maxPivotAngle);
    }
  }

  /**
   * @return {@code true} if the intake's current position is above the stowed threshold, {@code
   *     false} otherwise.
   */
  public boolean isStowed() {
    return pivotInputs.positionRadians
        >= JsonConstants.intakeConstants.stowThresholdAngle.in(Radians);
  }

  /**
   * Returns whether the intake pivot is high enough up that the hood needs to start/stay stowing so
   * that it doesn't tear the net.
   *
   * @return {@code true} if the hood should stow, {@code false} if it isn't in danger of tearing
   *     the net
   */
  public boolean shouldStartStowingHood() {
    return pivotInputs.positionRadians
        >= JsonConstants.intakeConstants.pivotStartStowingHoodAngle.in(Radians);
  }

  /**
   * Returns whether the intake pivot is high enough up that the turret needs to stop moving so that
   * it doesn't tear the net.
   *
   * @return {@code true} if the turret should stop, {@code false} if it isn't in danger of tearing
   *     the net
   */
  public boolean shouldStopTurret() {
    return pivotInputs.positionRadians
        >= JsonConstants.intakeConstants.pivotStopTurretAngle.in(Radians);
  }
}
